src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,235 @@
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+- `lpf` is a function returned by `make_logjoint`.
+
+Optional arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
+"""
+function mode_estimation(
+    model::DynamicPPL.Model, 
+    lpf; 
+    optim_options=Optim.Options(),
+    kwargs...
+)
+    # Do some initialization.
+    b = bijector(model)
+    binv = inv(b)
+
+    spl = DynamicPPL.SampleFromPrior()
+    vi = DynamicPPL.VarInfo(model)
+    init_params = model(vi, spl)
+    init_vals = vi[spl]
+
+    # Construct target function.
+    target(x) = lpf(x)
+    hess_target(x) = lpf(x; unlinked=true)
+
+    # Optimize!
+    M = Optim.optimize(target, init_vals, optim_options)
+
+    # Retrieve the estimated values.
+    vals = binv(M.minimizer)
+
+    # Get the VarInfo at the MLE/MAP point, and run the model to ensure 
+    # correct dimensionality.
+    vi[spl] = vals
+    model(vi) # XXX: Is this a necessary step?
+
+    # Make one transition to get the parameter names.
+    ts = [Turing.Inference.Transition(DynamicPPL.tonamedtuple(vi), DynamicPPL.getlogp(vi))]
+    varnames, _ = Turing.Inference._params_to_array(ts)
+
+    # Store the parameters and their names in an array.
+    vmat = NamedArrays.NamedArray(vals, varnames)
+
+    # Try to generate the information matrix.
+    try
+        # Calculate Hessian and information matrix.
+        info = ForwardDiff.hessian(hess_target, vals)
+        info = inv(info)
+        mat = NamedArrays.NamedArray(info, (varnames, varnames))
+
+        # Create the standard errors.
+        ses = sqrt.(diag(info))
+
+        # Calculate t-stats.
+        tstat = vals ./ ses
+
+        # Make a summary table.
+        stable = NamedArrays.NamedArray(
+            [vals ses tstat], 
+            (varnames, ["parameter", "std_err", "tstat"]))
+
+        # Return a wrapped-up table.
+        return ModeResult(vmat, mat, M, stable, M.minimum)
+    catch err
+        @warn "Could not compute Hessian matrix" err
+        stable = NamedArrays.NamedArray([vals repeat([missing], length(vals)) repeat([missing], length(vals))], (varnames, ["parameter", "std_err", "tstat"]))
+        return ModeResult(vmat, missing, M, stable, M.minimum)
+    end
+end
+
+"""
+    MLE(model::DynamicPPL.Model; kwargs...)
+
+Returns a maximum likelihood estimate of the given `model`.
+
+Arguments: 
+
+- `model` is a `DynamicPPL.Model`.
+
+Keyword arguments:
+
+- `optim_options` is a `Optim.Options` struct that allows you to change the number
+  of iterations run in an MLE estimate.
+
+Usage:
+
+```julia
+using Turing
+
+@model function f()
+    m ~ Normal(0, 1)
+    1.5 ~ Normal(m, 1)
+    2.0 ~ Normal(m, 1)
+end
+
+model = f()
+mle_estimate = MLE(model)
+
+# Manually setting the optimizers settings.
+mle_estimate = MLE(model, optim_options=Optim.Options(iterations=500))
+```
+"""
+function MLE(model::DynamicPPL.Model; kwargs...)
```

Comment:
IMO this should be
```suggestion
function mle(model::DynamicPPL.Model; kwargs...)
```
since `MLE` suggests incorrectly that an object of type `MLE` is returned.

Reply:
Alternatively, one could use singleton types `MLE` and `MAP` as inputs to `mode_estimation` instead of the cost function.